File: cpp/49_Group_Anagrams.cpp
/*
Given an array of strings, group anagrams together.

Example:
Input: ["eat", "tea", "tan", "ate", "nat", "bat"],
Output:
[
  ["ate","eat","tea"],
  ["nat","tan"],
  ["bat"]
]

Note:
All inputs will be in lowercase.
The order of your output does not matter.
*/
#include <iostream>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
    public:
        vector<vector<string>> groupAnagrams(vector<string>& strs) {
            vector<vector<string>> res;
            unordered_map<string, int> map;
            int len = strs.size(), begin = 0;
            string temp = "";
            for(int i = 0; i < len; i++){
                string ss = strs[i];
                temp = ss;
                sort(ss.begin(), ss.end());
                if(map.find(ss) != map.end()) {
                    res[map[ss]].push_back(temp);
                } else {
                    res.push_back({temp});
                    map[ss] = begin++;
                }
            }
            return res;
        }
};

int main() {
    Solution sol;
    return 0;
}

/*


*/

File: py/49_Group_Anagrams.py
"""
Given an array of strings, group anagrams together.

Example:
Input: ["eat", "tea", "tan", "ate", "nat", "bat"],
Output:
[
  ["ate","eat","tea"],
  ["nat","tan"],
  ["bat"]
]

Note:
All inputs will be in lowercase.
The order of your output does not matter.
"""
from typing import List

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        res = {}
        for word in strs:
            key = tuple(sorted(word))
            res[key] = res.get(key, []) + [word]
        return list(res.values())
